File: binary_search/find_first_and_last_position_of_element_in_sorted_array.cpp
//
// Created by Harsha Vardhan on 18/06/22.
//

#include <iostream>
#include <vector>

using namespace std;

enum occurence {
    FIRST,
    LAST
};

int binarySearch(vector<int> &nums, int target, occurence occur) {
    int l = 0, r = nums.size() - 1;
    int index = -1;
    while (l <= r) {
        int mid = l + (r - l) / 2;
        if (nums[mid] > target) {
            r = mid - 1;
        } else if (nums[mid] < target) {
            l = mid + 1;
        } else {
            index = mid;
            if (occur == FIRST) {
                r = mid - 1;
            } else {
                l = mid + 1;
            }
        }
    }
    return index;
}

vector<int> searchRange(vector<int> &nums, int target) {
    int first = binarySearch(nums, target, FIRST);
    int last = binarySearch(nums, target, LAST);
    vector<int> ans(2);
    ans[0] = first;
    ans[1] = last;
    return ans;
}

int main() {
    vector<int> nums{5, 7, 7, 8, 8, 10};
    int target = 8;
    searchRange(nums, target);
    return 0;
}

File: go/binary_search/find_first_and_last_position_of_element_in_sorted_array.go
package binary_search

func searchRange(nums []int, target int) []int {
	s, e := 0, len(nums)-1

	for s <= e {
		mid := s + (e-s)/2
		if nums[mid] == target {
			i := mid
			j := mid
			for i >= 0 && nums[i] == target {
				i--
			}
			for j < len(nums) && nums[j] == target {
				j++
			}
			return []int{i + 1, j - 1}
		} else if nums[mid] < target {
			s = mid + 1
		} else {
			e = mid - 1
		}
	}

	return []int{-1, -1}
}
